app/services/ai_backend/gemini.rb

```diff
@@ -0,0 +1,116 @@
+class AIBackend::Gemini < AIBackend
+  include Tools
+
+  # Rails system tests don't seem to allow mocking because the server and the
+  # test are in separate processes.
+  #
+  # In regular tests, mock this method or the TestClients::Gemini class to do
+  # what you want instead.
+  def self.client
+    if Rails.env.test?
+      ::TestClient::Gemini
+    else
+      ::Gemini
+    end
+  end
+
+  def initialize(user, assistant, conversation = nil, message = nil)
+    super(user, assistant, conversation, message)
+    begin
+      raise ::OpenAI::ConfigurationError if assistant.api_service.requires_token? && assistant.api_service.effective_token.blank?
+      Rails.logger.info "Connecting to Gemini API server at #{assistant.api_service.url} with access token of length #{assistant.api_service.effective_token.to_s.length}"
+      @client = self.class.client.new(credentials: {service: "generative-language-api",
+                                                    api_key: assistant.api_service.effective_token,
+                                                    version: "v1beta"},
+                                      options: { model: assistant.language_model.api_name,
+                                      server_sent_events: true })
+    rescue ::Faraday::UnauthorizedError => e
+      raise OpenAI::ConfigurationError
+    end
+  end
+
+  def client_method_name
+    :stream_generate_content
+  end
+
+  def configuration_error
+    ::OpenAI::ConfigurationError
+  end
+
+  def set_client_config(config)
+    super(config)
+
+    @client_config = {
+      contents: config[:messages],
+      system_instruction: ( system_message(config[:instructions]) if @assistant.language_model.supports_system_message?)
+    }.compact
+  end
+
+  def get_oneoff_message(instructions, messages, params = {})
+    set_client_config(
+      messages: preceding_conversation_messages,
+      instructions: full_instructions,
+    )
+
+    response = @client.send(client_method_name, @client_config)
+    response.dig("candidates",0,"content","parts",0,"text")
+  end
+
+  def stream_next_conversation_message(&chunk_handler)
+    set_client_config(
+      messages: preceding_conversation_messages,
+      instructions: full_instructions,
+    )
+
+    begin
+      if Rails.env.test?
+        @client.send(client_method_name, @client_config).each do |intermediate_response|
+          content_chunk = intermediate_response.dig("candidates",0,"content","parts",0,"text")
+          yield content_chunk if content_chunk != nil
+        end
+      else
+        response = @client.send(client_method_name, @client_config) do |intermediate_response, parsed, raw|
+          content_chunk = intermediate_response.dig("candidates",0,"content","parts",0,"text")
+          yield content_chunk if content_chunk != nil
+        end
```

Comment:
I put a bunch of breakpoints in place and this is where I have issue. Look at this output from my debug. Did you run into this?

```
=> 75:       binding.pry
    76:       response = @client.send(client_method_name, @client_config) do |intermediate_response, parsed, raw|
    77:         binding.pry
    78:         content_chunk = intermediate_response.dig("candidates",0,"content","parts",0,"text")
    79:         yield content_chunk if content_chunk != nil
    80:       end
    81:     end
    82:   rescue ::Faraday::UnauthorizedError => e
    83:     puts e.message
    84:     raise OpenAI::ConfigurationError
    85:   end
    86:   return nil
    87: end

[1] pry(#<AIBackend::Gemini>)> exit
objc[38948]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called.
objc[38948]: +[__NSCFConstantString initialize] may have been in progress in another thread when fork() was called. We cannot safely call it or ignore it in the fork() child process. Crashing instead. Set a breakpoint on objc_initializeAfterForkError to debug.
```

Reply:
Yes, I have that problem but I thought it was my pc setup. It is resolved with this line in bin/dev:

  export OBJC_DISABLE_INITIALIZE_FORK_SAFETY=YES

I got into the other stuff in this week. Thanks @krschacht !